### PR TITLE
new admin libraries page

### PR DESCRIPTION
**Pull request by imwhatiam**

### CODE REVIEW

File: static/scripts/sysadmin-app/views/libraries.js [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+        initialize: function() {
+            this.libraryCollection = new LibraryCollection();
+            this.listenTo(this.libraryCollection, 'add', this.addOne);
+            this.listenTo(this.libraryCollection, 'reset', this.reset);
+            this.render();
+        },
+
+        render: function() {
+            var data = {'cur_tab': 'all',};
```
Comment (freeplant) [inline static/scripts/sysadmin-app/views/libraries.js]: Remove ","

File: seahub/api2/endpoints/admin/devices.py [hunk 2]
```diff
@@ -57,7 +57,11 @@ def get(self, request, format=None):
             result['platform'] = device.platform
             return_results.append(result)
 
-        return Response(({'has_next_page': has_next_page}, return_results))
+        page_info = {
+            'has_next_page': has_next_page,
+            'current_page': current_page
+        }
+        return Response((page_info, return_results))
```
Comment (llj) [inline seahub/api2/endpoints/admin/devices.py]: m

File: seahub/api2/endpoints/admin/libraries.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+        for repo in repos_all:
+            repo_info = get_repo_info(repo)
+            return_results.append(repo_info)
+
+        page_info = {
+            'has_next_page': has_next_page,
+            'current_page': current_page
+        }
+        return Response((page_info, return_results))
```
Comment (llj) [inline seahub/api2/endpoints/admin/libraries.py]: m

File: seahub/api2/endpoints/admin/libraries.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    throttle_classes = (UserRateThrottle,)
+    permission_classes = (IsAdminUser,)
+
+    def delete(self, request, repo_id, format=None):
+        repo = seafile_api.get_repo(repo_id)
+        if not repo:
+            return Response({'success': True})
+
+        if get_system_default_repo_id() == repo_id:
+            error_msg = ('System library can not be deleted.')
```
Comment (llj) [inline seahub/api2/endpoints/admin/libraries.py]: `()` ?
 or `_()`?

File: seahub/api2/endpoints/admin/libraries.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+            return api_error(status.HTTP_400_BAD_REQUEST, error_msg)
+
+        if is_org_context(request):
+            org_id = seafserv_threaded_rpc.get_org_id_by_repo_id(repo_id)
+            usernames = get_related_users_by_org_repo(org_id, repo_id)
+            repo_owner = seafile_api.get_org_repo_owner(repo_id)
+        else:
+            org_id = -1
+            usernames = get_related_users_by_repo(repo_id)
```
Comment (llj) [inline seahub/api2/endpoints/admin/libraries.py]: ?

Comment (imwhatiam) [inline seahub/api2/endpoints/admin/libraries.py]: """When a repo is deleted, an event would be added to every user in all
        groups to which this repo is shared.

File: seahub/api2/endpoints/admin/devices.py (hunk 2 above)
Comment (imwhatiam) [inline seahub/api2/endpoints/admin/devices.py]: Most of the time, providing something like this in your response is sufficient for updating the pagination state.

[{"total_entries": 100}, [{}, {}, ...]]

File: seahub/api2/endpoints/admin/libraries.py (hunk 5 above)
Comment (imwhatiam) [inline seahub/api2/endpoints/admin/libraries.py]: do'not need to notice related users when delete a repo from admin

File: seahub/templates/js/sysadmin-templates.html [hunk 6]
```diff
@@ -12,8 +12,9 @@ <h3 class="hd">{% trans "System Admin" %}</h3>
         <li class="tab">
             <a href="{% url "sys_settings" %}"><span class="sf2-icon-cog2"></span>{% trans "Settings" %}</a>
         </li>
-        <li class="tab">
-            <a href="{{ SITE_ROOT }}sys/seafadmin/"><span class="sf2-icon-library"></span>{% trans "Libraries" %}</a>
+
+        <li class="tab <% if (cur_tab == 'libraries') { %> tab-cur<% } %>">
```
Comment (llj) [inline seahub/templates/js/sysadmin-templates.html]: `class="tab` 后面多了个空格

File: seahub/templates/js/sysadmin-templates.html [hunk 7]
```diff
@@ -234,3 +235,225 @@ <h2 class="alc">{% trans "No sync errors" %}</h2>
...
+
+<script type="text/template" id="library-item-tmpl">
+    <% if (encrypted) { %>
+    <td><img src="{{MEDIA_URL}}img/sync-folder-encrypt-20.png" title="{% trans "Encrypted"%}" alt="{% trans "directory icon" %}" /></td>
+    <% } else { %>
+    <td><img src="{{MEDIA_URL}}img/sync-folder-20.png?" title="{% trans "Read-Write"%}" alt="{% trans "directory icon" %}" /></td>
+    <% } %>
+    <% if (name) { %>
+        <% if (enable_sys_admin_view_repo && is_pro && !encrypted) { %>
+        <td><a href="#libraries/<%- id %>/dirents/"><%- name %></a></td>
```
Comment (llj) [inline seahub/templates/js/sysadmin-templates.html]: ?

File: seahub/templates/js/sysadmin-templates.html [hunk 8]
```diff
@@ -234,3 +235,225 @@ <h2 class="alc">{% trans "No sync errors" %}</h2>
...
+        <% } else { %>
+        <td><%- name %></td>
+        <% } %>
+    <% } else { %>
+    <td>--</td>
+    <% } %>
+    <td><%- file_count %> / <%- size_formatted %></td>
+    <td style="font-size:11px;"><%- id %></td>
```
Comment (llj) [inline seahub/templates/js/sysadmin-templates.html]: 上面两行，用 `=`

File: seahub/templates/js/sysadmin-templates.html [hunk 9]
```diff
@@ -234,3 +235,225 @@ <h2 class="alc">{% trans "No sync errors" %}</h2>
...
+    <td>
+        <% if (owner) { %>
+        <a href="{{ SITE_ROOT }}useradmin/info/<%- owner %>/"><%- owner %></a>
```
Comment (llj) [inline seahub/templates/js/sysadmin-templates.html]: url 里的那个 owner, 用 `encodeURIComponent` 处理。

File: seahub/templates/js/sysadmin-templates.html [hunk 10]
```diff
@@ -234,3 +235,225 @@ <h2 class="alc">{% trans "No sync errors" %}</h2>
...
+        <% } else { %>
+        --
+        <% } %>
+    </td>
+    <td>
+        <a href="#" class="sf2-icon-delete sf2-x repo-delete-btn op-icon vh" title="{% trans "Delete" %}" aria-label="{% trans "Delete" %}"></a>
+        <a href="#" class="sf2-icon-move sf2-x repo-transfer-btn op-icon vh" title="{% trans "Transfer" %}" aria-label="{% trans "Transfer" %}"></a>
```
Comment (llj) [inline seahub/templates/js/sysadmin-templates.html]: ?

**Comment by llj** [comment]
python 函数都加上注释吧，说明下这个函数是做什么的。

**Pull request merged**